src/logging.c

```diff
@@ -0,0 +1,55 @@
+#include <stdarg.h>
+#include <time.h>
+
+#include "logging.h"
+
+
+/******************************************************
+ *** LOCAL DEFINES                                  ***
+ ******************************************************/
+
+/******************************************************
+ *** LOCAL VARIABLES                                ***
+ ******************************************************/
+
+static LoggingLevel min_dbg_lvl = FATAL;
+
+/******************************************************
+ *** LOCAL METHODS                                  ***
+ ******************************************************/
+
+/******************************************************
+ *** EXPOSED METHODS                                ***
+ ******************************************************/
+
+void log_set_lvl(const CLIArguments* conf) {
+    min_dbg_lvl = (LoggingLevel) conf->debug;
+
+    // minimum debug level cannot be higher than fatal
+    if (min_dbg_lvl > FATAL)
+        min_dbg_lvl = FATAL;
+
+    // minimum debug level cannot be lower than debug
+    if (min_dbg_lvl < DEBUG)
+        min_dbg_lvl = DEBUG;
+
+    LOG_INFO("Log level initialized to %d.", min_dbg_lvl);
+}
+
+__attribute__((format(printf, 4, 5))) void log_str(LoggingLevel dbg_lvl, const char* dbg_lvl_str, FILE* stream, const char *msg, ...) {
+    if (dbg_lvl < min_dbg_lvl) return;
```

Comment:
Is this check really necessary? log_str _should_ only be called by the LOG_* macros with fixed levels, right?

Reply:
Yes it is necessary. Wouldn't it be for this check, the log level setup would have no effect.

Reply:
I'm not sure I understand. Let's assume we don't have this check, what would happen then?

Reply:
You could not select the minimum level a log message should have for displaying.

Reply:
Suppose you select that to be ERROR. If not for this check, all DEBUG, INFO and WARNING messages would be printed as well.